### ISSUE TITLE
GDScript 2.0: Infinite recursion crash in _guess_identifier_type

### ISSUE THREAD

**Issue by JohanAR** [godotengine/godot#63592]
### Godot version

v4.0.alpha12.official [2c11e6d9e]

### System information

Xubuntu 22.04

### Issue description

Editor crashes as soon as I type a " inside the brackets at d[]

```
extends MyBase

static func test() -> Dictionary:
	var d = super()
	# Type a " inside [] on the next line to crash Godot
	d[]
	return d
```

Parent class implements the same function:
```
class_name MyBase extends Node

static func test() -> Dictionary:
	return {
		"a": 1
	}
```

Here is the end of the stack trace:

```
.........
#4040 _guess_expression_type (p_context=..., p_expression=0x55557a02a9e0, r_type=...) at modules/gdscript/gdscript_editor.cpp:1377
#4041 0x0000555556c68ecb in _guess_method_return_type_from_base (r_type=..., p_method=..., p_base=..., p_context=...) at modules/gdscript/gdscript_editor.cpp:2265
#4042 _guess_expression_type (p_context=..., p_expression=0x55557a1b5700, r_type=...) at modules/gdscript/gdscript_editor.cpp:1601
#4043 0x0000555556c672e9 in _guess_identifier_type (r_type=..., p_identifier=..., p_context=...) at modules/gdscript/gdscript_editor.cpp:1910
#4044 _guess_expression_type (p_context=..., p_expression=0x55557a02a9e0, r_type=...) at modules/gdscript/gdscript_editor.cpp:1377
#4045 0x0000555556c68ecb in _guess_method_return_type_from_base (r_type=..., p_method=..., p_base=..., p_context=...) at modules/gdscript/gdscript_editor.cpp:2265
#4046 _guess_expression_type (p_context=..., p_expression=0x55557a1b5700, r_type=...) at modules/gdscript/gdscript_editor.cpp:1601
#4047 0x0000555556c672e9 in _guess_identifier_type (r_type=..., p_identifier=..., p_context=...) at modules/gdscript/gdscript_editor.cpp:1910
#4048 _guess_expression_type (p_context=..., p_expression=0x55557a02a9e0, r_type=...) at modules/gdscript/gdscript_editor.cpp:1377
#4049 0x0000555556c68ecb in _guess_method_return_type_from_base (r_type=..., p_method=..., p_base=..., p_context=...) at modules/gdscript/gdscript_editor.cpp:2265
#4050 _guess_expression_type (p_context=..., p_expression=0x55557a1b5700, r_type=...) at modules/gdscript/gdscript_editor.cpp:1601
#4051 0x0000555556c672e9 in _guess_identifier_type (r_type=..., p_identifier=..., p_context=...) at modules/gdscript/gdscript_editor.cpp:1910
#4052 _guess_expression_type (p_context=..., p_expression=0x55557a02a9e0, r_type=...) at modules/gdscript/gdscript_editor.cpp:1377
#4053 0x0000555556c68ecb in _guess_method_return_type_from_base (r_type=..., p_method=..., p_base=..., p_context=...) at modules/gdscript/gdscript_editor.cpp:2265
#4054 _guess_expression_type (p_context=..., p_expression=0x55557a1b5700, r_type=...) at modules/gdscript/gdscript_editor.cpp:1601
#4055 0x0000555556c672e9 in _guess_identifier_type (r_type=..., p_identifier=..., p_context=...) at modules/gdscript/gdscript_editor.cpp:1910
#4056 _guess_expression_type (p_context=..., p_expression=0x55557a02a330, r_type=...) at modules/gdscript/gdscript_editor.cpp:1377
#4057 0x0000555556c7766b in GDScriptLanguage::complete_code (this=<optimized out>, p_code=..., p_path=..., p_owner=0x0, r_options=0x7fffffffc428, r_forced=@0x7fffffffc427: false, r_call_hint=...)
    at modules/gdscript/gdscript_editor.cpp:2758
#4058 0x0000555557a233fa in ScriptTextEditor::_code_complete_script (this=0x5555766e56b0, p_code=..., r_options=0x7fffffffc428, r_force=@0x7fffffffc427: false)
    at editor/plugins/script_text_editor.cpp:736
#4059 0x0000555557db1620 in CodeTextEditor::_complete_request (this=0x55557675f740) at editor/code_editor.cpp:915
#4060 0x000055555a14d13b in Object::emit_signalp (this=this@entry=0x55557675e120, p_name=..., p_args=p_args@entry=0x0, p_argcount=p_argcount@entry=0) at core/object/object.cpp:1013
#4061 0x000055555830da6d in Object::emit_signal<>(StringName const&) (p_name=..., this=0x55557675e120) at ./core/object/object.h:821
#4062 CodeEdit::request_code_completion (this=0x55557675e120, p_force=<optimized out>) at scene/gui/code_edit.cpp:1862
#4063 0x000055555a14d13b in Object::emit_signalp (this=this@entry=0x5555760abd40, p_name=..., p_args=p_args@entry=0x0, p_argcount=p_argcount@entry=0) at core/object/object.cpp:1013
#4064 0x000055555827b931 in Object::emit_signal<>(StringName const&) (p_name=..., this=0x5555760abd40) at ./core/object/object.h:821
#4065 Timer::_notification (this=0x5555760abd40, p_what=<optimized out>) at scene/main/timer.cpp:60
#4066 0x000055555a141f88 in Object::notification (this=0x5555760abd40, p_notification=25, p_reversed=<optimized out>) at core/object/object.cpp:741
#4067 0x0000555558262a7f in SceneTree::_notify_group_pause (this=<optimized out>, p_group=..., p_notification=25) at scene/main/scene_tree.cpp:904
#4068 0x0000555558264bef in SceneTree::process (this=0x555560630f30, p_time=0.0069030000000000003) at scene/main/scene_tree.cpp:453
#4069 0x00005555560149f7 in Main::iteration () at main/main.cpp:2874
#4070 0x0000555555fded21 in OS_LinuxBSD::run (this=this@entry=0x7fffffffcbb0) at platform/linuxbsd/os_linuxbsd.cpp:537
#4071 0x0000555555fd0e49 in main (argc=<optimized out>, argv=0x7fffffffd0e8) at platform/linuxbsd/godot_linuxbsd.cpp:72
```

### Steps to reproduce

It crashes every time I load the project and try to complete that line

### Minimal reproduction project

[EditorCrashTest.zip](https://github.com/godotengine/godot/files/9213838/EditorCrashTest.zip)

**Comment by AntonioDell** [comment]
I investigated a bit and it seems to be exclusively an issue for inherited functions calling super() and trying to figure out, which type the super function returns. In contrast, calling a different static function from the parent, will not crash the editor and type inference works fine.
I extended @JohanAR s MRP to include this facet of 
the problem and will start to tackle the issue.

[63592_Test.zip](https://github.com/godotengine/godot/files/9216813/63592_Test.zip)
